sql: do not use leased table descriptors modified during a transaction

vivekmenezes opened the pull request:
This change makes a sql transaction cache table descriptors modified
by it and use them within the transaction without leasing the descriptors
themselves. This allows a transaction to consistently see its own changes.

The need for this change is described here:

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/table_descriptor_lease.md#accommodating-schema-changes-within-transactions

In the future, we plan to decouple the user transaction from the
transaction acquiring a lease. This change will prevent a
deadlock between an open transaction that has modified a descriptor
and the other transaction that is attempting to acquire a lease
on the descriptor. This change breaks the dependency between these
two transactions, by having the former not depend on the latter.

vivekmenezes committed:
sql: do not use leased table descriptors modified during a transaction

This change makes a sql transaction cache table descriptors modified
by it and use them within the transaction without leasing the descriptors
themselves. This allows a transaction to consistently see its own changes.

The need for this change is described here:

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/table_descriptor_lease.md#accommodating-schema-changes-within-transactions

In the future, we plan to decouple the user transaction from the
transaction acquiring a lease. This change will prevent a
deadlock between an open transaction that has modified a descriptor
and the other transaction that is attempting to acquire a lease
on the descriptor. This change breaks the dependency between these
two transactions, by having the former not depend on the latter.

related to #16727

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16988)
<!-- Reviewable:end -->

vivekmenezes commented:
making another change to this PR. hold on

vivekmenezes commented:
Ready for review

jordanlewis commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/table.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLlne2y6vD5Ft_YOJl:-KpLlne2y6vD5Ft_YOJm:b-r5tkjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L264)):*
> ```Go
> 	// with this TableCollection. This allows a transaction to see
> 	// its own modifications while bypassing the table lease mechanism.
> 	// The table lease mechanism will its own transaction to read
> ```

missing verb here

---

*[pkg/sql/table.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLm9dDF4FeFjzCGxRH:-KpLm9dDF4FeFjzCGxRI:b-1pdp7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L269)):*
> ```Go
> 	// TableCollection and invisible to other transactions. A dropped
> 	// table is marked dropped.
> 	uncommittedTables []sqlbase.TableDescriptor
> ```

When do these get cleared? Based on the comment on TableCollection, it seems like the table collection is session-scoped. But that would imply that it is never reset once created at the start of a session. Is that comment wrong? Otherwise it seems like these lists will just grow permanently and produce incorrect results.

---

*[pkg/sql/table.go, line 333 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLh8zJokzIylogNhlq:-KpLh8zJokzIylogNhlr:bwtuh4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L333)):*
> ```Go
> 	}
> 
> 	if dbID == 0 {
> ```

This could use a comment - I guess it's the case where we don't have a cached database?

---

*[pkg/sql/table.go, line 350 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLhbG9u55N7-uCubc-:-KpLhbG9u55N7-uCubc0:b1ygzfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L350)):*
> ```Go
> 	} else if table != nil {
> 		if log.V(2) {
> 			log.Infof(ctx, "found uncommitted table %d", table.ID)
> ```

Would this be better as an Eventf, so it can show up in traces?

---

*[pkg/sql/table.go, line 415 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLhehXjSkCjiH-NcjG:-KpLhehXjSkCjiH-NcjH:bcrt5hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L415)):*
> ```Go
> 		if table.ID == tableID {
> 			if log.V(2) {
> 				log.Infof(ctx, "found uncommitted table %d", tableID)
> ```

eventf?

---

*[pkg/sql/logictest/testdata/logic_test/rename_table, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLiQMwa9cxTVjz5Ohd:-KpLiQMwa9cxTVjz5Ohe:b10ca2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/logictest/testdata/logic_test/rename_table#L256)):*
> ```
> 
> statement error table \"d.kv2\" does not exist
> INSERT INTO d.kv2 (k,v) VALUES ('c', 'd')
> ```

Nice tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/table.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLlne2y6vD5Ft_YOJl:-KpQj2JxzhX1JKWhKmGB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L264)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

missing verb here
</blockquote></details>

Done.

---

*[pkg/sql/table.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLm9dDF4FeFjzCGxRH:-KpQkmVE3YO6-5qj7I8V:b-gi0cgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L269)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

When do these get cleared? Based on the comment on TableCollection, it seems like the table collection is session-scoped. But that would imply that it is never reset once created at the start of a session. Is that comment wrong? Otherwise it seems like these lists will just grow permanently and produce incorrect results.
</blockquote></details>

I've added a comment at the top about how releaseTables() is responsible for clearing the collection and when it is called.

---

*[pkg/sql/table.go, line 333 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLh8zJokzIylogNhlq:-KpQn-gNixQ1-xans9HU:b-l4396t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L333)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This could use a comment - I guess it's the case where we don't have a cached database?
</blockquote></details>

Added a comment

---

*[pkg/sql/table.go, line 350 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLhbG9u55N7-uCubc-:-KpQnusve0gvCM_LtlLf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L350)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Would this be better as an Eventf, so it can show up in traces?
</blockquote></details>

Done.

---

*[pkg/sql/table.go, line 415 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLhehXjSkCjiH-NcjG:-KpQnw6VVdOui8QqhIfh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/table.go#L415)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

eventf?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/rename_table, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-KpLiQMwa9cxTVjz5Ohd:-KpQo5w7-sqAJc2tCHZx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a19d053ea3f0806373475c75b7c74f5371ac49ac/pkg/sql/logictest/testdata/logic_test/rename_table#L256)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Nice tests.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@jordanlewis I rebased this change. Do you have any more comments on this? TFTR!

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16988#-:-Kpn56HGtq27GtOdiQtq:bnfp4nl)*
<!-- Sent from Reviewable.io -->
